### PR TITLE
Adds Observable support

### PR DESCRIPTION
**Pull request by danielwertheim**
This PR adds observable support as defined in #303

### CODE REVIEW

File: src/NATS.Client/Rx/INATSObservable.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace NATS.Client.Rx
+{
+    /// <summary>
+    /// Marker interface to have shipped extension methods to not interfear
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/INATSObservable.cs]:
```suggestion
    /// Marker interface to have shipped extension methods to not interfere
```

File: src/NATS.Client/Rx/INATSObservable.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    /// with offical RX extensions.
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/INATSObservable.cs]:
```suggestion
    /// with official RX extensions.
```

File: src/NATS.Client/Rx/NATSObservable.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        public IDisposable Subscribe(IObserver<T> observer)
+        {
+            if (observer == null)
+                throw new ArgumentNullException(nameof(observer));
+
+            var sub = new ObserverSubscription(observer, DisposeSubscription);
+            
+            if(!subscriptions.TryAdd(sub.Id, sub))
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/NATSObservable.cs:74]: Should we consider returning the existing subscription if this is called twice?

File: src/NATS.Client/Rx/Ops/DelegatingObserver.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        private readonly Action<T> onNext;
+        private readonly Action<Exception> onError;
+        private readonly Action onCompleted;
+
+        internal DelegatingObserver(
+            Action<T> onNext,
+            Action<Exception> onError = null,
+            Action onCompleted = null)
+        {
+            this.onNext = onNext;
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/Ops/DelegatingObserver.cs]:
```suggestion
            this.onNext = onNext ?? throw new ArgumentNullException(nameof(onNext));
```

File: src/NATS.Client/Rx/Ops/SafeObserver.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        private readonly Action<T> onNext;
+        private readonly Action<Exception> onError;
+        private readonly Action onCompleted;
+
+        internal SafeObserver(
+            Action<T> onNext,
+            Action<Exception> onError = null,
+            Action onCompleted = null)
+        {
+            this.onNext = onNext;
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/Ops/SafeObserver.cs]:
```suggestion
            this.onNext = onNext ?? throw new ArgumentNullException(nameof(onNext));
```

File: src/NATS.Client/Rx/NATSObservable.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        private sealed class ObserverSubscription : IDisposable
+        {
+            private readonly Action<ObserverSubscription> disposer;
+
+            public readonly int Id;
+            public readonly IObserver<T> Observer;
+
+            public ObserverSubscription(IObserver<T> observer, Action<ObserverSubscription> disposer)
+            {
+                Id = observer.GetHashCode();
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/NATSObservable.cs:89]: We're using hash codes for uniqueness, however, someone could conceptually make their own and our invariant would be broken. We're looking for unique-by-reference-equality, right?

File: src/NATS.Client/Rx/Ops/MapObservable.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        public MapObservable(INATSObservable<TSrc> src, Func<TSrc, TResult> predicate)
+        {
+            this.src = src ?? throw new ArgumentNullException(nameof(src));
+            this.mapper = predicate ?? throw new ArgumentNullException(nameof(predicate));
+        }
+
+        public void Dispose() => src?.Dispose();
+
+        public IDisposable Subscribe(IObserver<TResult> observer)
+            => src.SubscribeSafe(new MapObserver(observer, mapper));
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/Ops/MapObservable.cs:32]: Does this need a test for `observer` against `null`?

File: src/NATS.Client/Rx/Ops/WhereObservable.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        public WhereObservable(INATSObservable<T> src, Func<T, bool> predicate)
+        {
+            this.src = src ?? throw new ArgumentNullException(nameof(src));
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+        }
+
+        public void Dispose() => src?.Dispose();
+
+        public IDisposable Subscribe(IObserver<T> observer)
+            => src.SubscribeSafe(new WhereObserver(observer, predicate));
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/Ops/WhereObservable.cs:32]: Does this need a test for `observer` against `null`?

File: src/NATS.Client/Rx/RxExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        /// <summary>
+        /// Subscribes to the passed subject and returns a hot observable. Hence unless you
+        /// subscribe to the observable, no message will be handled and old messages will
+        /// not be delivered, only new messages.
+        /// </summary>
+        /// <param name="cn"></param>
+        /// <param name="subject"></param>
+        /// <returns></returns>
+        public static INATSObservable<Msg> Observe(this IConnection cn, string subject)
+            => cn.SubscribeAsync(subject).ToObservable();
```
Comment (watfordgnf) [inline src/NATS.Client/Rx/RxExtensions.cs]: We should probably test `cn` against `null`.

**Review by watfordgnf — COMMENTED** [review]
Some general comments, liking the way this looks!

File: src/NATS.Client/Rx/NATSObservable.cs (hunk 6 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/NATSObservable.cs:89]: Just as if we would add the observer to e.g. a HashSet. We could switch for a GUID if you feel it is that important?

File: src/NATS.Client/Rx/NATSObservable.cs (hunk 3 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/NATSObservable.cs:74]: But why should you call it twice?

File: src/NATS.Client/Rx/Ops/DelegatingObserver.cs (hunk 4 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/Ops/DelegatingObserver.cs]: Completely missed I could accept suggestion. It's been fixed though

File: src/NATS.Client/Rx/Ops/SafeObserver.cs (hunk 5 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/Ops/SafeObserver.cs]: Fixed

File: src/NATS.Client/Rx/NATSObservable.cs (hunk 3 above)
Comment (watfordgnf) [reply src/NATS.Client/Rx/NATSObservable.cs:74]: Perhaps this is a non-issue if users would never logically pass the same observer twice to Subscribe.

File: src/NATS.Client/Rx/Ops/MapObservable.cs (hunk 7 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/Ops/MapObservable.cs:32]: The `MapObserver` now requires it.

File: src/NATS.Client/Rx/Ops/WhereObservable.cs (hunk 8 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/Ops/WhereObservable.cs:32]: The WhereObserver now requires it.

File: src/NATS.Client/Rx/RxExtensions.cs (hunk 9 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/RxExtensions.cs]: We now check so that it isn't null.

File: src/NATS.Client/Rx/NATSObservable.cs (hunk 6 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/NATSObservable.cs:89]: The idea though, is that you should not add the exact same observer twice, hence e.g. a GUID would not work as I see it.

**Comment by danielwertheim** [comment]
This would also solve #149

File: src/NATS.Client/Rx/NATSObservable.cs (hunk 6 above)
Comment (watfordgnf) [reply src/NATS.Client/Rx/NATSObservable.cs:89]: What would be the downside to making the key simply the observer itself? We'd retain HashSet semantics and avoid the odd situation where two equal hash codes could occur for two distinct objects.

Comment (danielwertheim) [reply src/NATS.Client/Rx/NATSObservable.cs:89]: I guess it's just habit of as much as possible using a value object as a key due to allocations etc.

Using the "object" as the key will call it's GetHashCode when adding and GetHashCode + Equals e.g when getting so I'm not sure how that would solve the issue you want to solve.

File: src/NATS.Client/Rx/INATSObservable.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright 2015-2018 The NATS Authors
```
Comment (ColinSullivan1) [inline src/NATS.Client/Rx/INATSObservable.cs]: Minor thing...  Could you update the license headers to be `Copyright 2019 The NATS Authors`?

File: src/NATS.Client/Rx/NATSObservable.cs [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        {
+            foreach (var sub in subscriptions.Values)
+            {
+                try
+                {
+                    sub.Observer.OnNext(data);
+                }
+                catch
+                {
+                    //Missbehaving observer will remove its subscription
```
Comment (ColinSullivan1) [inline src/NATS.Client/Rx/NATSObservable.cs]: Typo `Misbehaving`.

File: src/NATS.Client/Rx/NATSObservableSubscription.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    public sealed class NATSObservableSubscription : NATSObservable<Msg>, IDisposable
+    {
+        private readonly IAsyncSubscription subscription;
+
+        private NATSObservableSubscription(IAsyncSubscription subscription)
+        {
+            this.subscription = subscription ?? throw new ArgumentNullException(nameof(subscription));
+
+            this.subscription.MessageHandler += OnIncomingMessage;
+            this.subscription.Start();
```
Comment (ColinSullivan1) [inline src/NATS.Client/Rx/NATSObservableSubscription.cs:27]: I'm wondering if it'd be useful (or feasible) to also setup an AsyncErrorHandler (normally set in [options](https://github.com/nats-io/nats.net/blob/master/src/NATS.Client/Options.cs#L72)) to intercept async errors from the server and invoke `onError` in the observable.  I know these errors are attached to a connection today, but it might be useful to surface slow consumer errors through `onError`.  wdyt?

File: src/Tests/IntegrationTests/TestRx.cs [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+                        sync.WaitForAll();
+                        Assert.Equal(payload1, interceptingOb1.OnNextResults.Single());
+                        Assert.Equal(payload2, interceptingOb2.OnNextResults.Single());
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public void WhenDispoingAnObserverItShouldNotReceiveMoreMessages()
```
Comment (ColinSullivan1) [inline src/Tests/IntegrationTests/TestRx.cs]: typo...

File: src/Tests/IntegrationTests/TestRx.cs [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+                        Assert.Equal(new[] { payload1.Data, payload2.Data }, interceptingOb1.OnNextResults.Select(i => i.Data).ToArray());
+                        Assert.Equal(new[] { payload1.Data }, interceptingOb2.OnNextResults.Select(i => i.Data).ToArray());
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public void WhenDispoingAnObservableNoMoreDispatchesShouldBeDone()
```
Comment (ColinSullivan1) [inline src/Tests/IntegrationTests/TestRx.cs]: Typo...

**Review by ColinSullivan1 — CHANGES_REQUESTED** [review]
This looks great!  Some minor comments/typos and a question regarding slow consumers.  Could you add some brief example code into the README?

It doesn't need to be part of this PR (can be separate), but an example project users could reference would go a long way to helping them started with this API.

File: src/NATS.Client/Rx/NATSObservableSubscription.cs (hunk 12 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/NATSObservableSubscription.cs:27]: Calling `OnError` would terminate the observer, meaning if the connection comes up again, the subscribed observer would not get any further dispatches. I don't think that is optimal.

File: src/NATS.Client/Rx/INATSObservable.cs (hunk 10 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/INATSObservable.cs]: Updated the ones added in this PR.

File: src/Tests/IntegrationTests/TestRx.cs (hunk 14 above)
Comment (danielwertheim) [reply src/Tests/IntegrationTests/TestRx.cs]: Fixed

File: src/Tests/IntegrationTests/TestRx.cs (hunk 13 above)
Comment (danielwertheim) [reply src/Tests/IntegrationTests/TestRx.cs]: Fixed

File: src/NATS.Client/Rx/NATSObservable.cs (hunk 11 above)
Comment (danielwertheim) [reply src/NATS.Client/Rx/NATSObservable.cs]: Fixed

**Comment by danielwertheim** [comment]
@ColinSullivan1 Added `RxSample` project under Samples and added code to README,

File: src/NATS.Client/Rx/NATSObservableSubscription.cs (hunk 12 above)
Comment (ColinSullivan1) [reply src/NATS.Client/Rx/NATSObservableSubscription.cs:27]: Ahh, good point.  I agree - thanks.

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.  Thanks - well done!

**Comment by ColinSullivan1** [comment]
I think this is ready to be merged.  @watfordgnf , wdyt?

**Review by watfordgnf — APPROVED** [review]
